syntax = "proto3";

import "common.proto";
import "ssid.proto";
import "sslcls.proto";
import "ssgwgs.proto";
import "ssgwio.proto";
import "sspcpx.proto";
import "ssdcds.proto";

message SSPkgHead
{
	enum MSGTYPE
	{
		NORMAL = 0;
		RPCREQ = 1;
		RPCRSP = 2;
	}

	enum PROXYTYPE
	{
		END = 0;
		FORWARD = 1;
		BROADCAST = 2;
	}

	enum LOGICTYPE
	{
		BOTH = 0;
		CPP = 1;
		LUA = 2;
	}

	optional NODETYPE from_node_type = 1;
	optional uint32 from_node_id = 2;
	optional NODETYPE to_node_type = 3;
	optional uint32 to_node_id = 4;
	optional SSID id = 5;
	optional MSGTYPE msg_type = 6;
	optional uint64 rpc_id = 7;
	optional PROXYTYPE proxy_type = 8;
	optional LOGICTYPE logic_type = 9;
}

message SSPkgBody
{
	oneof body
	{
		SSLCLSPkgBody lcls_body = 1;
		SSGWGSPkgBody gwgs_body = 2;
		SSGWIOPkgBody gwio_body = 3;
		SSPCPXPkgBody pcpx_body = 4;
		SSDCDSPkgBody dcds_body = 5;
	}
}

message SSPkg
{
	SSPkgHead head = 1;
	SSPkgBody body = 2;
}